File: js-project/src/main2.js
/**
 * callback関数になっているライブラリをasync/awaitにしたい場合の例
 */
const main = async () => {
    console.log('[JavaScript] main start')
    const result = await testFunc()
    console.log(result)
    console.log('[JavaScript] main end')
}

// promiseでcallbackをasync/awaitで実行できるように、callbackで帰ってくるAPIを包んでます
const testFunc = () => {
    return new Promise(resolve => {
        console.log("testFunc method")
        callbackMethod(ret => {
            resolve(ret)
        })
    })
}

// callbackになっているライブラリの変わり
const callbackMethod = callback => {
    callback(456)
}

main()

File: kotlin-project/app/src/main/java/sample/kotlin/Main2.kt
/**
 * callback関数になっているライブラリをasync/awaitにしたい場合の例
 */

package sample.kotlin

import kotlinx.coroutines.*
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

fun main() {
    println("[kotlin] main start")

    // JavaScriptでいうと、awaitっぽい感じ。なの通りでrunBlocking内の処理が終わるまで待ってくれる
    runBlocking {
        val func = async { testFunc() }
        val result = func.await()
        println(result)
    }
    println("[kotlin] main end")
}

/**
 * JavaScriptでいうところのnew Promise()と同じ様な効果。
 * suspendCoroutineでcallbackをasync/awaitで実行できるように、callbackで帰ってくるAPIを包んでます
 */
suspend fun testFunc(): Int {
    println("testFunc method")
    return suspendCoroutine { cont ->
        callbackMethod {
            cont.resume(it)
        }
    }
}

// callbackになっているライブラリの変わり
fun callbackMethod(ret: (Int) -> Unit) {
    ret(456)
}
